### PR TITLE
Added Firmware v9 GNSS configuration

### LINKED ISSUE #197: Cannot configure multi-GNSS for ZED F9P
**Describe the bug**
Configuration of the F9P to use any other GNSS system besides GPS results in a 'UBX-ACK-NAK' error message.

**To Reproduce**
1. Checkout 'master'
2. Configure the ublox package to use this config [zed_f9p-nmea.txt](https://github.com/KumarRobotics/ublox/files/10413713/zed_f9p-nmea.txt) (unfortunately github does not support yaml extension, I've changed it to .txt which has to be changed to yaml before initialization)
3. Make sure the F9P receiver is connected to `/dev/ttyACM0` or adapt the config file to your needs.
4. Launch the ublox node and receive the following error msg when e.g. :
Example output with `gnss/glonass: true`, which has been set in the provided F9P config
image

**Expected behavior**
One of both would be appreciated (for more information see [link](https://cdn.sparkfun.com/assets/learn_tutorials/8/5/6/ZED-F9P_UBX_NMEA_and_RTCM_protocols.pdf) chapter: 5.9.8.1 GNSS system configuration):
- The configuration of multiple GNSS systems for the F9P should work based on the current approach (needs to be adjusted to `UBX-CFGVALSET, UBX-CFG-VALGET, UBX-CFG-VALDEL`)
- The configuration of multiple GNSS systems for the F9P should be adjusted to use `CFG-SIGNAL-???`

**Desktop:**
 - OS: Ubuntu 20.04
 - ROS Version: ROS1 Noetic

**Additional context**
I'm working on a project using the C099-F9P board to localize a robot. So far I was able to set up the NMEA/RTCM messages in ROS. The communication with the ntrip client works fine and I'm able to acquire float-RTK. I want to further increase the accuracy with additional satellite sources, currently I'm achieving 127mm hacc and 165mm vacc.

### PR DESCRIPTION
**Pull request by brandonbeggs**
Now uses CfgVALSET for GNSS configuration.  Should resolve #197 & #147

### CODE REVIEW

File: ublox_msgs/msg/CfgVALSETCfgdata.msg [hunk 1]
```diff
@@ -5,3 +5,28 @@ uint32 key      # Key for configuration item being changed
...
+uint32 BDS_B2_ENABLE =      271646734   #0x1031000e
+
+uint32 QZSS_ENABLE =        271646756   #0x10310024
+uint32 QZSS_L1CA_ENABLE =   271646738   #0x10310012
+uint32 QZSS_L1S_ENABLE =    271646740   #0x10310014
+uint32 QZSS_L2C_ENABLE =    271646741   #0x10310015
+
+uint32 GLO_ENABLE =         271646757   #0x10310025
+uint32 GLO_L1_ENABLE =      271646744   #0x10310018
+uint32 GLO_L2_ENABLE =      271646746   #0x1031001a
```
Comment (clalancette) [inline ublox_msgs/msg/CfgVALSETCfgdata.msg]: Can you please move these to the top of the file?   Also, a space between the comment `#` and the hex number would be appreciated on each line.

File: ublox_gps/src/node.cpp [hunk 2]
```diff
@@ -701,6 +701,21 @@ void UbloxNode::processMonVer() {
             }
             continue;
           }
+
+          // u-blox F9 modules support additional positioning signals
+          if(strs[0] == "MOD"){
```
Comment (clalancette) [inline ublox_gps/src/node.cpp]:
```suggestion
          // u-blox F9 modules support additional positioning signals
          else if (strs[0] == "MOD")
          {
```

File: ublox_gps/src/node.cpp [hunk 3]
```diff
@@ -701,6 +701,21 @@ void UbloxNode::processMonVer() {
...
+            std::vector<std::string> moduleField;
+            moduleField = stringSplit(strs[1], "-");
+            if(moduleField.size() > 1){
```
Comment (clalancette) [inline ublox_gps/src/node.cpp]:
```suggestion
            if (moduleField.size() > 1)
            {
```

File: ublox_gps/src/node.cpp [hunk 4]
```diff
@@ -701,6 +701,21 @@ void UbloxNode::processMonVer() {
...
+              if(moduleField[1].substr(0,2) == "F9"){
```
Comment (clalancette) [inline ublox_gps/src/node.cpp]:
```suggestion
              if (moduleField[1].substr(0,2) == "F9")
              {
```

File: ublox_gps/include/ublox_gps/ublox_firmware9.hpp [hunk 5]
```diff
@@ -15,12 +18,28 @@ namespace ublox_node {
...
+    * flash memory based on the ROS parameters.
+    */
+  bool configureUblox(std::shared_ptr<ublox_gps::Gps> gps) override;
+private:
+  /**
+    * @brief Populate the CfgVALSETCfgData data type
+    *
+    * @details A helper function used to generate a configuration for a singal signal. 
+    */
+  ublox_msgs::msg::CfgVALSETCfgdata generateSignalConfig_(uint32_t signalID, bool enable);
```
Comment (clalancette) [inline ublox_gps/include/ublox_gps/ublox_firmware9.hpp]:
```suggestion
  ublox_msgs::msg::CfgVALSETCfgdata generateSignalConfig(uint32_t signalID, bool enable);
```

File: ublox_gps/include/ublox_gps/ublox_firmware9.hpp [hunk 6]
```diff
@@ -15,12 +18,28 @@ namespace ublox_node {
...
+    * @brief Configure settings specific to firmware 9 based on ROS parameters.
+    *
+    * @details Configure GNSS.  The hardware has internal logic for
+    * detecting differences between the new and active GNSS
+    * confifuration and will internally trigger a reset if necessary.
+    * Configure the NMEA if desired by the user. It also may clear the
+    * flash memory based on the ROS parameters.
+    */
+  bool configureUblox(std::shared_ptr<ublox_gps::Gps> gps) override;
+private:
```
Comment (clalancette) [inline ublox_gps/include/ublox_gps/ublox_firmware9.hpp]:
```suggestion
  bool configureUblox(std::shared_ptr<ublox_gps::Gps> gps) override;

private:
```

File: ublox_gps/src/ublox_firmware9.cpp [hunk 7]
```diff
@@ -16,4 +17,113 @@ UbloxFirmware9::UbloxFirmware9(const std::string & frame_id, std::shared_ptr<dia
...
+  if (gnss_->isSupported("GLO_L2"))
+  {
+    signalCfgData.push_back(generateSignalConfig_(signal::GLO_L2_ENABLE, enable_glonass_));
+  }
+
+  cfg_signal.cfgdata.resize(signalCfgData.size());
+  for (size_t i = 0; i < signalCfgData.size(); ++i)
+  {
+    cfg_signal.cfgdata[i] = signalCfgData[i];
+  }
```
Comment (clalancette) [inline ublox_gps/src/ublox_firmware9.cpp]: Do we need to do this copy?  That is, can we just `push_back` onto `cfg_signal.cfgdata` vector as we need to above, and remove `signalCfgData` completely?

File: ublox_gps/include/ublox_gps/ublox_firmware8.hpp [hunk 8]
```diff
@@ -75,6 +75,7 @@ class UbloxFirmware8 : public UbloxFirmware7Plus<ublox_msgs::msg::NavPVT> {
   bool clear_bbr_{false};
   bool save_on_shutdown_{false};
 
+ private:
```
Comment (clalancette) [inline ublox_gps/include/ublox_gps/ublox_firmware8.hpp]:
```suggestion
private:
```

**Review by clalancette — CHANGES_REQUESTED** [review]

File: ublox_msgs/msg/CfgVALSETCfgdata.msg (hunk 1 above)
Comment (brandonbeggs) [reply ublox_msgs/msg/CfgVALSETCfgdata.msg]: It has now been moved to the very top assuming you wanted it above the "# See Cfg-VALSET" comment.

File: ublox_gps/src/ublox_firmware9.cpp (hunk 7 above)
Comment (brandonbeggs) [reply ublox_gps/src/ublox_firmware9.cpp]: Ah this is why we do code reviews.  I was oblivious that `cfg_signal.cfgdata` was a `std::vector`.  Should be fixed now.  This clean things up a bit.

File: ublox_gps/include/ublox_gps/ublox_firmware9.hpp [hunk 9]
```diff
@@ -15,12 +18,29 @@ namespace ublox_node {
...
 class UbloxFirmware9 final : public UbloxFirmware8 {
 public:
   explicit UbloxFirmware9(const std::string & frame_id, std::shared_ptr<diagnostic_updater::Updater> updater, std::shared_ptr<FixDiagnostic> freq_diag, std::shared_ptr<Gnss> gnss, rclcpp::Node* node);
+
+  /**
+    * @brief Configure settings specific to firmware 9 based on ROS parameters.
+    *
+    * @details Configure GNSS.  The hardware has internal logic for
+    * detecting differences between the new and active GNSS
+    * confifuration and will internally trigger a reset if necessary.
```
Comment (clalancette) [inline ublox_gps/include/ublox_gps/ublox_firmware9.hpp]:
```suggestion
    * configuration and will internally trigger a reset if necessary.
```

File: ublox_gps/include/ublox_gps/ublox_firmware9.hpp [hunk 10]
```diff
@@ -15,12 +18,29 @@ namespace ublox_node {
...
+    * Configure the NMEA if desired by the user. It also may clear the
+    * flash memory based on the ROS parameters.
+    */
+  bool configureUblox(std::shared_ptr<ublox_gps::Gps> gps) override;
+
+private:
+  /**
+    * @brief Populate the CfgVALSETCfgData data type
+    *
+    * @details A helper function used to generate a configuration for a singal signal.
```
Comment (clalancette) [inline ublox_gps/include/ublox_gps/ublox_firmware9.hpp]:
```suggestion
    * @details A helper function used to generate a configuration for a single signal. 
```

**Review by clalancette — CHANGES_REQUESTED** [review]

**Review by clalancette — APPROVED** [review]
Thank you for all of the iteration.  This looks good to me now, I'll go ahead and merge.

**Pull request merged**